### DISCUSSION TITLE
Error requesting VendorDirectFulfillmentInventory().submit_inventory_update

### DISCUSSION THREAD

**Discussion by kprkpr** [saleweaver/python-amazon-sp-api#835] [FAQ]
**Describe the bug**
Hi, I'm trying to update the inventory of my VendorCentral DF account, and when try to do it, it returns a "Invalid content was found starting with element 'marketplaceIds'. One of '{inventory}' is expected."

**To Reproduce**

(See the code attached for more)
The call is:
`
inv = VendorDirectFulfillmentInventory(marketplace=Marketplaces.ES,credentials=credentials).submit_inventory_update(warehouseId='BSVM',inventory=jsonl)
print(inv)
`

It returns:
`
Traceback (most recent call last):
  File "/root/Desktop/python3file.py", line 43, in <module>
    inv = VendorDirectFulfillmentInventory(marketplace=Marketplaces.ES,credentials=credentials).submit_inventory_update(warehouseId='BSVM',inventory=jsonl)
  File "/usr/local/lib/python3.10/dist-packages/sp_api/base/helpers.py", line 19, in wrapper
    return function(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/sp_api/api/vendor_direct_fulfillment_inventory/vendor_direct_fulfillment_inventory.py", line 58, in submit_inventory_update
    return self._request(fill_query_params(kwargs.pop('path'), warehouseId), data=kwargs)
  File "/usr/local/lib/python3.10/dist-packages/sp_api/base/client.py", line 153, in _request
    return self._check_response(res, res_no_data, bulk, wrap_list)
  File "/usr/local/lib/python3.10/dist-packages/sp_api/base/client.py", line 176, in _check_response
    raise exception(error, headers=res.headers)
sp_api.base.exceptions.SellingApiBadRequestException: [{'code': 'InvalidInput', 'message': "Invalid content was found starting with element 'marketplaceIds'. One of '{inventory}' is expected.", 'details': ''}, {'code': 'InvalidInput', 'message': 'MarketplaceIds', 'details': ''}]
`

**Expected behavior**
Update the inventory of this article..

**Desktop (please complete the following information):**

 - OS: Windows, python 3.10.8, sp_api 0.17.5

**Additional context**

I tried to use a proxy to see the data sent and it's: (Removed indentificable information)

`
POST /vendor/directFulfillment/inventory/v1/warehouses/BSVM/items HTTP/2
Host: sellingpartnerapi-eu.amazon.com
User-Agent: python-sp-api-0.17.5
Accept-Encoding: gzip, deflate
Accept: */*
Connection: keep-alive
X-Amz-Access-Token: REMOVED
X-Amz-Date: 20221212T091249Z
Content-Type: application/json
Content-Length: 289
Authorization: REMOVED
{"inventory": {"sellingParty": {"partyId": "REMOVED"}, "isFullUpdate": false, "items": [{"buyerProductIdentifier": "REMOVED", "availableQuantity": {"amount": 100, "unitOfMeasure": "Each"}, "isObsolete": false}]}, "marketplaceIds": ["REMOVED"], "MarketplaceIds": ["REMOVED"]}
`

Other API calls like "VendorDirectFulfillmentOrders().get_orders()" works, then I assume that credentials are not the problem

[EXAMPLE.TXT](https://github.com/saleweaver/python-amazon-sp-api/files/10206488/EXAMPLE.TXT)

**Comment by saleweaver** [comment]
Fix your input data, the error message even tells you where it’s wrong.

**Reply by kprkpr** [reply]
Yes, I read that the error says like why there is a marketplaceIds (That it's two times in the sent data by the way), but I don't know why it's the marketplaceIds in the sent data in the first place, because I don't added this, I only added the inventory json..
If it's because is in VendorDirectFulfillmentInventory(**marketplace=Marketplaces.ES**.... If i delete it, amazon says access denied "[{'message': 'Access to requested resource is denied.', 'code': 'Unauthorized', 'details': ''}]" and it connects to  sellingpartnerapi-na.amazon.com instead of sellingpartnerapi-eu.amazon.com

**Comment by kprkpr** [comment]
Hi
I continue facing this error..
My code is like this:
![image](https://user-images.githubusercontent.com/5132053/212280754-be77fc63-fdd8-47d1-b013-b1a5a5a1ebf0.png)
And I don't know why the sent request by the library is:
![image](https://user-images.githubusercontent.com/5132053/212280983-2343000e-559c-4b41-9b1f-b7eec6798c0c.png)
If you can help me I will appreciate a lot..

**Comment by kprkpr** [comment]
No solution? 😣 Continues to not work in various update methods. In the "VendorDirectFulfillmentShipping(marketplace=Marketplaces.ES,credentials=credentials).submit_shipment_status_updates(shipmentStatusUpdates=jsonl)" is the same.. It sends same "marketplaceIds" but I don't put anywhere...
![image](https://user-images.githubusercontent.com/5132053/223727234-7de8817d-774a-4945-a46c-26673d676602.png)
![image](https://user-images.githubusercontent.com/5132053/223727408-bd1077d9-2f9d-4ee3-9f9c-1ed058ee2629.png)

**Comment by bjquinn** [comment]
Hello, I'm getting the same error under the same circumstances, my only difference is I'm not passing marketplace=Marketplaces.ES, as I'm US and it defaults to that.  Does submit_inventory_update just not work at all?  Has anyone gotten this to work?

**Comment by bjquinn** [comment]
Ok, for anyone else who has an issue with this, as far as I can tell, the problem is that this particular API call does NOT want the MarketplaceIds or marketplaceIds parameters passed to it.  There's actually a mechanism within python-amazon-sp-api to refrain from sending the marketplace ID parameter.  This is used for lots of other API calls, including several other ones in the Vendor Direct Fulfillment family.  At any rate, it's missing from VendorDirectFulfillmentInventory.

A workaround that worked for me is by modifying <site-packages>/sp_api/api/vendor_direct_fulfillment_inventory/vendor_direct_fulfillment_inventory.py.  Change the last line of that file to the following:

`return self._request(fill_query_params(kwargs.pop('path'), warehouseId), data=kwargs, add_marketplace=False)`

I'm going to file a new bug and reference this issue/discussion.
